### ISSUE TITLE
Black window when refocusing some fullscreen application (generally Steam games)

### ISSUE THREAD

**Issue by Syudagye** [leftwm/leftwm#712]
**Describe the bug**
Switching to a tag where an application is launched full-screen can make it appear black and interactions with this window are impossible.
You can still use the window manager, but not the application.
The application do not crash, it is just impossible to interact with it.
This seem to be happening with certain games, generally steam games.

**To Reproduce**
1. Start the application full-screen on a tag
2. Switch to another tag
3. Go back to the first tag
4. Boom, can't play the game anymore :(

**Expected behavior**
Being able to refocus the window correctly and not have a full-screen black window with no interactions possible.

**Screenshots**
[here](https://youtu.be/z53fUhxkGD8) a little recording.

**Environment:**
 - OS/Distro: Arch Linux
 - Output of `leftwm-check`:
```
:: LeftWM version: 0.3.0
:: LeftWM git hash: c002da7
:: Loading configuration . . .
    -> Configuration loaded OK
:: Checking keybinds . . .
    -> All keybinds OK
:: Checking environment . . .
    -> Environment OK
:: Checking theme . . .
    -> Theme OK
```

**Additional context**
Application that had this issue for me:
- Rivals or Aether
- Geometry Dash

[osu](https://aur.archlinux.org/packages/osu) used to have the same issue a while ago, but now it seems like it's fixed

**Comment by Syudagye** [comment]
yes, i think it's only steam games running with proton. I tested native games and there was no issue at all (at least with those i tested)

**Comment by VuiMuich** [comment]
Just to be clear 'native' as in steam games natively built for linux or purely native to linux?
Might we narrow this down on being `proton`'s fault?

**Comment by Syudagye** [comment]
I think we can say it's proton's fault, maybe even wine, but i don't have any non working wine games outside of steam so i can't say anything about that.
Since I've seen this behavior only in leftwm, it may be related to the way fullscreen focus is handled, or focus in general ?

<sub>
ps: I forgot about this issue, sorry for the late response
</sub>

**Comment by Syudagye** [comment]
I don't think it's related to graphics drivers because it's working fine in other WMs. (I tested with i3, but I remember not having issues with XFCE). It may be related to the way leftwm gives focus to full-screen windows when you switch to their tag.

**Comment by hertg** [comment]
A similar issue was reported in #718, which I closed for now. It was just flickering and not the window staying black though.
@Syudagye Does the issue still occur? Maybe even in other games since you opened the issue? What graphics card are you using?

I think we need a reproducer to further investigate this. I'm leaving it open for now, but we might close it if no other users report such an issue and no further progress is made to pin the issue to LeftWM in the coming months.

**Comment by Syudagye** [comment]
Yes this is still happening, still on certain fullscreen applications. I have also another game that starts as it should but about 2 seconds in it just becomes a black screen too. It seems to fall in the same category as #718 (kind of?).
I recently acquired a laptop and i have the same issue on it.
For graphics cards i have a Radeon RX580 for my Desktop and a GTX 1650 Mobile for my laptop.

I'll try to reproduce #718 on both my desktop and my laptop to see if i get the same result.

**Comment by Syudagye** [comment]
Update for this issue:

I tried finding the cause of this issue, but didn't find anything.
I thought i was related to giving the focus to the window, so I investigated here. I found that the `_NET_WM_STATE_FOCUSED` wasn't given to the window when it is focused, so i implemented that quickly but nothing changed.
I looked at other xlib based window managers' implementation of the focus change, and they are all doing the same thing as leftwm, I can't find any differences (all of them use `XSetInputFocus`).

Then i found out that [this](https://github.com/leftwm/leftwm/blob/main/leftwm-core/src/display_servers/xlib_display_server/xwrap/window.rs#L292) sends the `WM_TAKE_FOCUS` atom to the window, but it seems to not be called on the windows I'm interested in. So i made it so it sends it anyway, even if it is not supported by the window. But it didn't had any effect either.

__In short:__ The focus implementation seems good, but it is not behaving as expected. Either there is a detail I am missing, or there is an issue elsewhere.

Also the windows affected by this issue seems to be programs (mainly games) launched through Wine/Proton and set to `fullscreen` mode. I've tested it with non-steam games an found the same issue. Setting those to a `borderless` mode fixes the issue, but not all of them have this option.

**Comment by VuiMuich** [comment]
Wild guess: are you using a compositor like picom?

**Comment by Syudagye** [comment]
yes, but this is not the cause. I usually start games without a compositor, and during my testing i tried with and without the compositor active, and the issue is still here.

**Comment by VuiMuich** [comment]
Would have been nice to have a culprit :stuck_out_tongue:

**Comment by VuiMuich** [comment]
@Syudagye do you know if this is specific to `leftwm`, or could this be reproduced in `dwm` or `bspwm` as well?

**Comment by Syudagye** [comment]
I think i have tried it in `dwm`, but i don't remember what it did, i'll check that as soon as i can.

**Comment by Syudagye** [comment]
Well I have tested with my unpatched dwm, and it has the same behavior. That sounds like a step forward on this investigation !

**Comment by Syudagye** [comment]
I found a workaround for this: enabling wine virtual desktop

To do so you need `winecfg`:
```bash
# replace $STEAM with your steam root and $APPID with the id of the game
# You can find the id when going to the store page in a web browser
WINEPREFIX=$STEAM/steamapps/compatdata/$APPID/pfx/ winecfg
```
Then go to "Display" and enable "Emulate virtual desktop". Don't forget to change the size of the desktop to the resolution of your display (for me it's 1920x1080 for example)

This is still a workaround, to the issue is not fixed yet, but this can avoid switching desktop for just doing one thing

**Comment by aeosynth** [comment]
i have this issue with a windows game ([vivid/stasis](https://store.steampowered.com/app/2093940/vividstasis/, free). i can also reproduce with the `ToggleFullScreen` command, in addition to switching tags.

**Comment by avlec** [comment]
I am having this happen with Helldivers 2 recently. The keyboard input doesn't work after the first key but the mouse does. I haven't played in 6 months and now it consistently happens unless I put the game in Windowed or Borderless Windowed. It also happens on Big Walk without black screening just the keyboard input breaks but that is resolved by flipping workspaces?
